### CONTEXT AND RELEVANCE

#### Which might answer these questions:
- Are there any potential biases in the data?
- If there are biases, what are the risks and implications?


In [23]:
# import library
import pandas as pd

# Relative path to read the CSV file from the "data" folder
df = pd.read_csv(r"./data/kalayaan/inventories.csv")

# Convert columns to their appropriate data types
df = df.infer_objects()

---

### DATA QUALITY

### Check for missing values

In [24]:
# Check for missing values
null_counts = df.isnull().sum()

print(f"{null_counts}\n-----------\n")

for column, count in null_counts.items():
    if count > 0:
        print(f"{column}: {count} null values")

_id                                    0
accessionId                            0
amoNumber                           3322
assetCategory                          0
assetDescription                       0
assetQuantity                          0
modeOfAcquisition                      0
acquisitionDate                     3319
acquisitionPrice                    3321
currentMarketValue                  3322
assetLocation                          0
roomNumber                             0
boxNumber                           1186
status                                 0
createdBy                              0
remarks                                8
assetNotes                          2073
createdAt                              0
updatedAt                           2969
__v                                 3257
borrowedAssets[0].date              3178
borrowedAssets[1].date              3320
borrowedAssets[0].borrowedBy        3178
borrowedAssets[1].borrowedBy        3320
borrowedAssets[0

### Check for duplicates

In [25]:
print("There are", f"\033[1m{df.duplicated().sum()}\033[0m", "duplicates found")

There are 0 duplicates found


---

### DATA STRUCTURE AND TYPES

In [26]:
rows, columns = df.shape
print(f"Number of rows: {rows:,}")
print(f"Number of columns: {columns:,}")

Number of rows: 3,322
Number of columns: 47


In [27]:
# Print the data types of the columns
print(df.dtypes)

_id                                  object
accessionId                          object
amoNumber                           float64
assetCategory                        object
assetDescription                     object
assetQuantity                         int64
modeOfAcquisition                    object
acquisitionDate                      object
acquisitionPrice                    float64
currentMarketValue                  float64
assetLocation                        object
roomNumber                           object
boxNumber                           float64
status                               object
createdBy                            object
remarks                              object
assetNotes                           object
createdAt                            object
updatedAt                            object
__v                                 float64
borrowedAssets[0].date               object
borrowedAssets[1].date               object
borrowedAssets[0].borrowedBy    

In [28]:
print("Count of columns in each type:\n")
df.dtypes.value_counts()

Count of columns in each type:



object     37
float64     9
int64       1
Name: count, dtype: int64

---

### OUTLIERS

In [29]:
def find_extremes(df, column):
    min_value = df[column].min()
    max_value = df[column].max()
    return min_value, max_value

In [30]:
min_val, max_val = find_extremes(df, 'assetQuantity')
print(f"Minimum: {min_val}, Maximum: {max_val}")

Minimum: -1, Maximum: 61


In [31]:
# The 5 smallest values
smallest_values = df['assetQuantity'].nsmallest(5)
print("Smallest values:\n", smallest_values)

# The 5 largest values
largest_values = df['assetQuantity'].nlargest(5)
print("Largest values:\n", largest_values)

Smallest values:
 1208   -1
529     0
912     0
1801    0
0       1
Name: assetQuantity, dtype: int64
Largest values:
 1833    61
279     49
1532    42
1032    40
1693    27
Name: assetQuantity, dtype: int64


---

### DATA DISTRIBUTIONS AND SUMMARY STATISTICS

In [32]:
import pandas as pd

def calculate_statistics(df, column):
    """
    Returns the mean, median, and mode for a specified column in the DataFrame.

    Parameters:
        df (pd.DataFrame): The DataFrame containing the data.
        column (str): The column name to calculate statistics for.

    Returns:
        dict: A dictionary containing the mean, median, and mode of the column.
    """
    mean_value = df[column].mean()
    median_value = df[column].median()
    mode_value = df[column].mode().iloc[0] if not df[column].mode().empty else None

    return {
        'mean': mean_value,
        'median': median_value,
        'mode': mode_value
    }

In [33]:
stats = calculate_statistics(df, 'assetQuantity')

# Formatting the output
formatted_stats = (
    f"Statistics for 'assetQuantity' column:\n"
    f"Mean: {stats['mean']}\n"
    f"Median: {stats['median']}\n"
    f"Mode: {stats['mode']}"
)

print(formatted_stats)

Statistics for 'assetQuantity' column:
Mean: 1.1881396748946418
Median: 1.0
Mode: 1


#### In simpler terms:

1. **Average Value**: The average assetQuantity is about 1.1881396748946418
2. **Middle Value**: The middle assetQuantity (when all assetQuantity are sorted) is 1.0
3. **Most Common Value**: The most frequently occurring assetQuantity is 1.

#### Interpretation:
- **Mostly Balanced**: The assetQuantity are mostly balanced around the middle value (1.1881).
- **Consistent Data**: Most assetQuantity are close to each other, with a slight tendency towards lower assetQuantity.